dm/pkg/checker/privilege.go

```diff
@@ -122,89 +143,214 @@ func (pc *SourceReplicatePrivilegeChecker) Name() string {
+							continue
+						}
+						delete(lackPriv[priv][dbName], tableName)
+						if len(lackPriv[priv][dbName]) == 0 {
+							delete(lackPriv[priv], dbName)
+						}
+						if len(lackPriv[priv]) == 0 {
+							delete(lackPriv, priv)
+						}
 					}
-				} else {
-					// currently, only SELECT privilege goes here. we didn't require SELECT to be granted globally,
-					// dumpling could report error if an allow-list table is lack of privilege.
-					// we only check that SELECT is granted on all columns, otherwise we can't SHOW CREATE TABLE
-					if len(privElem.Cols) == 0 {
-						delete(lackGrants, privElem.Priv)
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv]; !ok {
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv][dbName]; !ok {
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv][dbName][tableName]; !ok {
+					continue
+				}
+				// dumpling could report error if an allow-list table is lack of privilege.
+				// we only check that SELECT is granted on all columns, otherwise we can't SHOW CREATE TABLE
+				if privElem.Priv == mysql.SelectPriv && len(privElem.Cols) != 0 {
+					continue
+				}
+				delete(lackPriv[privElem.Priv][dbName], tableName)
+				if len(lackPriv[privElem.Priv][dbName]) == 0 {
+					delete(lackPriv[privElem.Priv], dbName)
+				}
+				if len(lackPriv[privElem.Priv]) == 0 {
+					delete(lackPriv, privElem.Priv)
+				}
+			}
+		}
+	}
+
+	if len(lackPriv) != 0 {
+		var buffer bytes.Buffer
+		// generate error message, for example
+		// lack of privilege1: {tableID1, tableID2, ...};lack of privilege2...
+		for p, tableMap := range lackPriv {
+			buffer.WriteString("lack of ")
+			buffer.WriteString(mysql.Priv2Str[p])
+			buffer.WriteString(" privilege")
+			if len(tableMap) != 0 {
+				buffer.WriteString(": {")
+			}
+			i := 0
+			for schema, tables := range tableMap {
+				if len(tables) == 0 {
+					buffer.WriteString(dbutil.ColumnName(schema))
+				}
+				j := 0
+				for table := range tables {
+					buffer.WriteString(dbutil.TableName(schema, table))
+					j++
+					if j != len(tables) {
+						buffer.WriteString(", ")
 					}
 				}
+				i++
+				if i != len(tableMap) {
+					buffer.WriteString("; ")
+				}
+			}
+			if len(tableMap) != 0 {
+				buffer.WriteString("}")
 			}
+			buffer.WriteString(";")
 		}
+		privileges := buffer.String()
+		result.Instruction = "You need grant related privileges."
+		log.L().Info("lack privilege", zap.String("err msg", privileges))
+		return NewError(privileges)
 	}
+	return nil
+}
 
-	if len(lackGrants) != 0 {
-		lackGrantsStr := make([]string, 0, len(lackGrants))
-		for g := range lackGrants {
-			lackGrantsStr = append(lackGrantsStr, mysql.Priv2Str[g])
+// checkTables map schema => {table1, table2, ...}.
+// lackPriv map privilege => schema => table.
+func genExpectPriv(privileges map[mysql.PrivilegeType]struct{}, checkTables map[string][]string) map[mysql.PrivilegeType]map[string]map[string]struct{} {
```

Comment:
after read the `verifyPrivileges`  i finnally know that the complex map is just a `Set`, i thinks this poor expression of golang can be imporved after Generics released........

at least, we need some thing like Set 😅